log: rate limiting followups

Crypt-iQ opened the pull request:
Followups to #32604.

There are two behavior changes:
- prefixing with `[*]` is done to all logs (regardless of `should_ratelimit`) per [this comment](https://github.com/bitcoin/bitcoin/pull/32604#discussion_r2195710943).
- a DEBUG_ONLY `-disableratelimitlogging` flag is added by default to functional tests so they don't encounter rate limiting.

Crypt-iQ committed:
test: remove noexcept(false) comment in ~DebugLogHelper

Crypt-iQ committed:
log: remove const qualifier from arguments in LogPrintFormatInternal

Co-Authored-By: l0rinc <pap.lorinc@gmail.com>

Crypt-iQ committed:
log: clarify RATELIMIT_MAX_BYTES comment, use RATELIMIT_WINDOW

Co-Authored-By: stickies-v <stickies-v@protonmail.com>

Crypt-iQ committed:
log: change LogLimitStats to struct LogRateLimiter::Stats

Clean up the noisy LogLimitStats and remove references to the time
window.

Co-Authored-By: stickies-v <stickies-v@protonmail.com>

Crypt-iQ committed:
log: clean up LogPrintStr_ and Reset, prefix all logs with "[*]" when there are suppressions

In LogPrintStr_:
- remove an unnecessary BCLog since we are in the BCLog namespace.
- remove an unnecessary \n when rate limiting is triggered since
  FormatLogStrInPlace will add it.
- move the ratelimit bool into an else if block.
- prefix all log lines with [*] when suppressions exist. Previously this
  was only done if should_ratelimit was true.

In Reset:
- remove an unnecessary \n since FormatLogStrInPlace will add it.
- Change Level::Info to Level::Warning.

stickies-v committed:
log: make m_limiter a shared_ptr

This allows us to safely and explicitly manage the dual dependency
on the limiter: one for the Logger, and one for the CScheduler.

stickies-v committed:
test: add ReadDebugLogLines helper function

Deduplicates repeated usage of the same functionality.

stickies-v committed:
test: logging_filesize_rate_limit improvements

- Add helper functions and structs to improve readability and
  reusability of test code
- Make tests more specific by comparing all produced log lines with
  expected log lines instead of relying on approximations or proxies.

stickies-v reviewed (commented):
code lgtm 8aa2726f1cfbc1e4267a796cb000c4eeae5910f5 - but would like to see if we can make improvements on the [`m_limiter` lifetime guarantees](https://github.com/bitcoin/bitcoin/pull/33011#discussion_r2226684372), and on the [`logging_filesize_rate_limit` tests](https://github.com/bitcoin/bitcoin/pull/33011#discussion_r2226805223). Both are not essential, but I think it could make sense to include them here.

Crypt-iQ committed:
log: avoid double hashing in SourceLocationHasher

Co-Authored-By: l0rinc <pap.lorinc@gmail.com>

Crypt-iQ commented on src/test/logging_tests.cpp:
```diff
@@ -307,13 +307,15 @@ BOOST_AUTO_TEST_CASE(logging_log_rate_limiter)
-    auto source_loc_1{std::source_location::current()};
-    auto source_loc_2{std::source_location::current()};
+    std::source_location source_loc_1{std::source_location::current()};
+    std::source_location source_loc_2{std::source_location::current()};
```
Fixed in 526403df23a2db781709e4494da3a9f79284531d

Crypt-iQ commented on src/test/logging_tests.cpp:
```diff
@@ -348,122 +350,137 @@ BOOST_AUTO_TEST_CASE(logging_log_rate_limiter)
+namespace {
+
+enum class Location {
+    FIRST,
```
Fixed in 526403df23a2db781709e4494da3a9f79284531d, renamed `DEBUG` to `DEBUG_LOG` since the former is a macro

Crypt-iQ commented on src/test/logging_tests.cpp:
```diff
@@ -348,122 +350,137 @@ BOOST_AUTO_TEST_CASE(logging_log_rate_limiter)
+    bool prev_log_threadnames{LogInstance().m_log_threadnames};
     LogInstance().m_log_threadnames = false;
 
+    int64_t line_length{1023};
```
Fixed in 526403df23a2db781709e4494da3a9f79284531d

Crypt-iQ commented on src/test/logging_tests.cpp:
```diff
@@ -348,122 +350,137 @@ BOOST_AUTO_TEST_CASE(logging_log_rate_limiter)
     log_file_size = GetFileSize(log_path);
     {
         ASSERT_DEBUG_LOG("Restarting logging");
         MockForwardAndSync(scheduler, 1min);
```
Fixed in 526403df23a2db781709e4494da3a9f79284531d

Crypt-iQ commented on src/test/logging_tests.cpp:
```diff
@@ -348,122 +350,137 @@ BOOST_AUTO_TEST_CASE(logging_log_rate_limiter)
-
     LogInstance().m_log_timestamps = prev_log_timestamps;
     LogInstance().m_log_sourcelocations = prev_log_sourcelocations;
     LogInstance().m_log_threadnames = prev_log_threadnames;
```
Fixed in 526403df23a2db781709e4494da3a9f79284531d

Crypt-iQ commented on src/test/logging_tests.cpp:
Implemented in 526403df23a2db781709e4494da3a9f79284531d, but I think there is a CI failure now.

Crypt-iQ commented:
Latest push 526403df23a2db781709e4494da3a9f79284531d -> d434155db5f9c34b8ab3e19038d824d161b34195 modifies `logging_filesize_rate_limit` to scan `ReadDebugLogLines` for the `[warning] Restarting logging` string in one place. This is hacky and is needed to make the CI pass when `DEBUG_LOCKCONTENTION` is specified.

The failure happens because:
- `Reset` is in the task queue of the scheduler
- `MockForwardAndSync` calls `MockForward`
- `Reset` will be called and will schedule another `Reset`. This will lock `newTaskMutex`.
- The `scheduleFromNow` call in `MockForwardAndSync` will lock `newTaskMutex`.

If there's lock contention here, it will trigger the failure. The logs will look like:
```
[14:38:33.794] [*] [warning] Excessive logging detected from test/logging_tests.cpp:392 (void logging_tests::(anonymous namespace)::LogFromLocation(Location, const std::string &)): >1048576 bytes logged during the last time window of 20s. Suppressing logging to disk from this source location until time window resets. Console logging unaffected. Last log entry.
[14:38:33.794] [*] a
[14:38:33.794] [*] b
[14:38:33.794] [*] c
[14:38:33.794] [lock] Enter: lock contention newTaskMutex, scheduler.cpp:74 started
[14:38:33.794] [warning] Restarting logging from test/logging_tests.cpp:392 (void logging_tests::(anonymous namespace)::LogFromLocation(Location, const std::string &)): 4 bytes were dropped during the last 20s.
[14:38:33.794] [lock] Enter: lock contention newTaskMutex, scheduler.cpp:74 completed (10μs)
[14:38:33.794] [lock] Enter: lock contention newTaskMutex, scheduler.cpp:74 started
[14:38:33.794] [lock] Enter: lock contention newTaskMutex, scheduler.cpp:74 completed (36μs)
[14:38:33.794] test/logging_tests.cpp(469): [1;31;49merror: in "logging_tests/logging_filesize_rate_limit": check ReadDebugLogLines().back().starts_with("[warning] Restarting logging") has failed[0;39;49m
```

stickies-v committed:
test: don't leak log category mask across tests

This ensures log tests behave consistently when other tests modify
the log category mask.

stickies-v commented on src/test/logging_tests.cpp:
```diff
@@ -337,138 +348,138 @@ BOOST_AUTO_TEST_CASE(logging_log_rate_limiter)
-        ASSERT_DEBUG_LOG("Restarting logging");
-        MockForwardAndSync(scheduler, 1min);
+        scheduler.MockForwardAndSync(time_window);
+        auto log_lines{ReadDebugLogLines()};
+        auto restart_pred = [](std::string& s) { return s.starts_with("[warning] Restarting logging"); };
+        BOOST_CHECK(std::any_of(log_lines.begin(), log_lines.end(), restart_pred));
```
Re https://github.com/bitcoin/bitcoin/pull/33011#issuecomment-3136494016:

Nice find re the lock contention. I was able to hit this from other locations in the test too on my machine. As you suggested offline, I think the best fix is to disable the LOCK category for this test. It might be helpful to first add a commit to prevent leaking categories across tests (e.g. https://github.com/stickies-v/bitcoin/commit/3d96ff75f161419654b14a7e9fd884e52aec26c4), and then the below diff should work?

<details>
<summary>git diff on 3d96ff75f1</summary>

```diff
diff --git a/src/test/logging_tests.cpp b/src/test/logging_tests.cpp
index 919222f5f8..df00d2e570 100644
--- a/src/test/logging_tests.cpp
+++ b/src/test/logging_tests.cpp
@@ -451,6 +451,12 @@ BOOST_FIXTURE_TEST_CASE(logging_filesize_rate_limit, LogSetup)
     LogInstance().m_log_sourcelocations = false;
     LogInstance().m_log_threadnames = false;
 
+// This test's scheduler schedules new tasks on 2 threads, potentially causing (non-circular)
+// lock contentions. Disable those logs so we can maintain a tight accounting.
+#ifdef DEBUG_LOCKCONTENTION
+    LogInstance().DisableCategory(BCLog::LogFlags::LOCK);
+#endif
+
     constexpr int64_t line_length{1024};
     constexpr int64_t num_lines{1024};
     constexpr int64_t bytes_quota{line_length * num_lines};
@@ -470,9 +476,7 @@ BOOST_FIXTURE_TEST_CASE(logging_filesize_rate_limit, LogSetup)
     TestLogFromLocation(Location::INFO_2, "c", Status::UNSUPPRESSED, /*suppressions_active=*/true);
     {
         scheduler.MockForwardAndSync(time_window);
-        auto log_lines{ReadDebugLogLines()};
-        auto restart_pred = [](std::string& s) { return s.starts_with("[warning] Restarting logging"); };
-        BOOST_CHECK(std::any_of(log_lines.begin(), log_lines.end(), restart_pred));
+        BOOST_CHECK(ReadDebugLogLines().back().starts_with("[warning] Restarting logging"));
     }
     // Check that logging from previously suppressed location is unsuppressed again.
     TestLogFromLocation(Location::INFO_1, log_message, Status::UNSUPPRESSED, /*suppressions_active=*/false);

```
</details>

Crypt-iQ replied:
Yup, I had also hit the lock contention logging issue in other parts of the test. I ran the suggested diff 100 times without any error. Will add.

stickies-v replied:
Yeah sorry that should've been `LogInstance().DisableCategory(BCLog::LogFlags::ALL);`. I still think that's worth keeping, so we can always assume `LogTest` starts without any categories (as is default)?

Crypt-iQ replied:
Per offline discussion, removed the `#ifdef DEBUG_LOCKCONTENTION` since the category masks are now properly reset between iterations (thanks!).

Crypt-iQ commented:
Latest push d434155db5f9c34b8ab3e19038d824d161b34195 -> f57496e:
- ensures the category mask does not leak across tests
- modifies `logging_filesize_rate_limit` to disable `BCLog::LogFlags::LOCK` if `DEBUG_LOCKCONTENTION` is defined

This gets rid of the test flake without hacking around `ReadDebugLogLines`.

Crypt-iQ commented on src/test/logging_tests.cpp:
```diff
@@ -367,106 +382,113 @@ BOOST_AUTO_TEST_CASE(logging_log_limit_stats)
+    case Location::DEBUG_LOG:
+        LogDebug(BCLog::LogFlags::HTTP, "%s\n", message);
+        return;
+    case Location::INFO_NOLIMIT:
```
Oh, I did not know functional tests were hitting the limit. I will add an option to disable the rate limiting.

Crypt-iQ committed:
config: add DEBUG_ONLY -logratelimit

Use -nologratelimit by default in functional tests if the bitcoind
version supports it.

Co-Authored-By: stickies-v <stickies-v@protonmail.com>

Crypt-iQ commented on test/functional/test_framework/test_node.py:
```diff
@@ -117,6 +117,7 @@ def __init__(self, i, datadir_path, *, chain, rpchost, timewait, timeout_factor,
             "-debugexclude=leveldb",
             "-debugexclude=rand",
             "-uacomment=testnode%d" % i,  # required for subversion uniqueness across peers
+            "-disableratelimitlogging",
```
thanks, will fix.

Crypt-iQ commented:
Latest push 31ef2b6 -> 8dd5c0a:
- adds a DEBUG_ONLY flag to disable the rate limiting. This was pointed out [here](https://github.com/bitcoin/bitcoin/pull/33011#discussion_r2246301270) that some of the functional tests were having their logs suppressed during normal operation.
- resets logging categories between tests without relying on disabling the `LOCK` category.

Should the DEBUG_ONLY flag get a release note?

stickies-v commented on src/init/common.cpp:
```diff
@@ -40,6 +40,7 @@ void AddLoggingArgs(ArgsManager& argsman)
     argsman.AddArg("-loglevelalways", strprintf("Always prepend a category and level (default: %u)", DEFAULT_LOGLEVELALWAYS), ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-printtoconsole", "Send trace/debug info to console (default: 1 when no -daemon. To disable logging to file, set -nodebuglogfile)", ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-shrinkdebugfile", "Shrink debug.log file on client startup (default: 1 when no -debug)", ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-disableratelimitlogging", "Disable rate limit logging", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
```
Probably good to add this to release notes?

<details>
<summary>git diff on 8dd5c0a044</summary>

```diff
diff --git a/doc/release-notes-32604.md b/doc/release-notes-32604.md
index d4d3726e86..1aaebc185b 100644
--- a/doc/release-notes-32604.md
+++ b/doc/release-notes-32604.md
@@ -4,7 +4,8 @@ Unconditional logging to disk is now rate limited by giving each source location
 a quota of 1MiB per hour. Unconditional logging is any logging with a log level
 higher than debug, that is `info`, `warning`, and `error`. All logs will be
 prefixed with `[*]` if there is at least one source location that is currently
-being suppressed. (#32604)
+being suppressed. Ratelimiting can be disabled globally with
+`-nologratelimit`. (#32604)
 
 When `-logsourcelocations` is enabled, the log output now contains the entire
 function signature instead of just the function name. (#32604)

```
</details>

Crypt-iQ replied:
Added

stickies-v replied:
_(sorry, should have edited my previous comment instead of deleting it, I misunderstood what you were replying to)_

You're right that we shouldn't advertise it. I thought adding it to release notes made sense for downstream projects that e.g. have their own test suite that could break because of this. I'm happy for the commit to be dropped, or be updated to highlight that it's a DEBUG_ONLY option.

stickies-v reviewed (approved):
ACK 8dd5c0a0447bbb6fe597aabfa725b397276da166

It's a debug-only option, so I suppose not technically a blocker, but would be best to [change](https://github.com/bitcoin/bitcoin/pull/33011#discussion_r2257061642) the `-disableratelimitlogging` option before merging.

Crypt-iQ commented on src/init.cpp:
```diff
@@ -1381,10 +1381,12 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
-        [&scheduler](auto func, auto window) { scheduler.scheduleEvery(std::move(func), window); },
-        BCLog::RATELIMIT_MAX_BYTES,
-        1h));
+    if (!args.IsArgSet("-disableratelimitlogging")) {
```
Thanks, implemented. I was trying to avoid adding a constant to `src/logging.h`.

maflcko reviewed (approved):
nothing blocking, just very minor nits/questions.

review ACK 7bc28668cc791d28fe824476d2c1f236c5a772d0 🕙

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK 7bc28668cc791d28fe824476d2c1f236c5a772d0 🕙
uUQiEQGUzj9jy99YEHObbOge+6ZfW5SIo6syrKomp3j0n6SYixr7vX43b6jXDN4DOO63UXBKS0C0plRGNiuaCQ==
```

</details>

Crypt-iQ commented on src/test/logging_tests.cpp:
```diff
@@ -394,6 +399,7 @@ BOOST_FIXTURE_TEST_CASE(logging_filesize_rate_limit, LogSetup)
     LogInstance().m_log_sourcelocations = false;
     bool prev_log_threadnames = LogInstance().m_log_threadnames;
     LogInstance().m_log_threadnames = false;
+    LogInstance().EnableCategory(BCLog::LogFlags::HTTP);
```
Oh ok, yup will fix.

Crypt-iQ replied:
Moved to the proper commit (fb9c7dd4ff7786bfb64aa6c174227e9dfdf04f64).

Crypt-iQ commented on src/test/logging_tests.cpp:
```diff
@@ -367,106 +382,107 @@ BOOST_AUTO_TEST_CASE(logging_log_limit_stats)
+    using Status = BCLog::LogRateLimiter::Status;
+    if (!suppressions_active) assert(status == Status::UNSUPPRESSED); // developer error
+
+    std::ofstream ofs(LogInstance().m_file_path, std::ios::out | std::ios::trunc); // clear debug log
```
Here's my guess why this works (may be wrong!):
- `std::ofstream` truncates the file since `std::ios::trunc` is used, this doesn't need a flush. Note that this doesn't reset the file offset of `m_fileout` (see [manpage for truncate](https://www.man7.org/linux/man-pages/man3/truncate.3p.html)).
- Next `LogFromLocation` is called. Note two things... 1) the file was opened in [append mode](https://github.com/bitcoin/bitcoin/blob/f679bad6052ad4b7ea0b57a406adc617746d54ed/src/logging.cpp#L63), and 2) buffering is [disabled](https://github.com/bitcoin/bitcoin/blob/f679bad6052ad4b7ea0b57a406adc617746d54ed/src/logging.cpp#L68). Since the file was opened in append mode, each write will move the file offset to the end of the file (see the part about O_APPEND in the manpage for the `write` syscall [here](https://www.man7.org/linux/man-pages/man2/write.2.html))
- Since there's no buffering, the `ReadDebugLogLines` call will correctly read the bytes written in `LogFromLocation`.

Crypt-iQ commented on src/logging.cpp:
```diff
@@ -549,26 +549,25 @@ void BCLog::LogRateLimiter::Reset()
-        m_available_bytes = 0;
+    if (bytes > available_bytes) {
+        dropped_bytes += bytes;
+        available_bytes = 0;
```
Fixed in 3c7cae49b692bb6bf5cae5ee23479091bed0b8be

achow101 commented:
ACK f1d25abcaa6a85fec08ea760036f2c7c02945ba8

l0rinc reviewed (approved):
I have done a quick scan over the commits only, left nits and questions, nothing critical.

Lightweight code review ACK f1d25abcaa6a85fec08ea760036f2c7c02945ba8

Note: I understand that this likely isn't consensus critical change but I would have preferred modifying the code once, in the original PR. Or maybe adding cleanup commits in a separate PR before the change. But if we need 12 follow-up commits, I think it tells us that the original one might have been merged a bit too quickly.

stickies-v commented on src/logging.cpp:
```diff
@@ -371,23 +371,32 @@ static size_t MemUsage(const BCLog::Logger::BufferedLog& buflog)
+    SchedulerFunction scheduler_func, uint64_t max_bytes, std::chrono::seconds reset_window)
 {
-    scheduler_func([this] { Reset(); }, reset_window);
+    auto limiter{std::shared_ptr<LogRateLimiter>(new LogRateLimiter(max_bytes, reset_window))};
```
I think the current approach is not exception safe wrt memory leak, but given that there are no complex ctors/allocators involved here and the function is called only sporadically, I think it's still the best alternative? How would you implement the `make_shared` approach?

stickies-v commented on src/logging.h:
```diff
@@ -104,47 +105,34 @@ namespace BCLog {
-    //! Keeps track of an individual source location and how many available bytes are left for logging from it.
-    class LogLimitStats
+    //! Fixed window rate limiter for logging.
+    class LogRateLimiter : public std::enable_shared_from_this<LogRateLimiter>
```
sorry, `std::enable_shared_from_this<LogRateLimiter>` was from an earlier approach, can be removed now
```suggestion
    class LogRateLimiter
```

Crypt-iQ replied:
Fixed in 3d630c2544e19480268426cda245796d4ce34ac3, nice job spotting this.

Crypt-iQ commented on src/logging.cpp:
```diff
@@ -371,23 +371,32 @@ static size_t MemUsage(const BCLog::Logger::BufferedLog& buflog)
+            shared_limiter->Reset();
+        }
+    },
+                   limiter->m_reset_window);
```
Will change in next push for readability.

Crypt-iQ commented on src/logging.cpp:
```diff
@@ -371,23 +371,32 @@ static size_t MemUsage(const BCLog::Logger::BufferedLog& buflog)
+    : m_max_bytes{max_bytes}, m_reset_window{reset_window} {}
+
+std::shared_ptr<BCLog::LogRateLimiter> BCLog::LogRateLimiter::Create(
+    SchedulerFunction scheduler_func, uint64_t max_bytes, std::chrono::seconds reset_window)
```
Thanks, will change in next push.

Crypt-iQ replied:
This is addressed in 3d630c2544e19480268426cda245796d4ce34ac3 by changing `BCLog::LogRateLimiter::Create` to take `SchedulerFunction&&`.

Crypt-iQ commented on src/logging.h:
```diff
@@ -104,47 +105,34 @@ namespace BCLog {
+        struct Stats {
+            //! Remaining bytes
+            uint64_t m_available_bytes;
+            //! Number of bytes that were consumed but didn't fit in the available bytes.
```
I think this is a bit more clear, resolving.

Crypt-iQ commented:
The latest push f1d25ab -> 5c74a0b:
- removes the release note for the DEBUG_ONLY `-logratelimit`
- addresses some minor comments about `LogRateLimiter` in 3d630c2544e19480268426cda245796d4ce34ac3

stickies-v commented:
re-ACK 5c74a0b397cb3db94761bad78801eed4544155b9

l0rinc commented on src/test/logging_tests.cpp:
```diff
@@ -294,22 +291,40 @@ BOOST_FIXTURE_TEST_CASE(logging_Conf, LogSetup)
-    auto sched_func = [&scheduler](auto func, auto window) { scheduler.scheduleEvery(std::move(func), window); };
-    BCLog::LogRateLimiter limiter{sched_func, max_bytes, reset_window};
+    ScopedScheduler scheduler{};
+    auto limiter_{scheduler.GetLimiter(max_bytes, reset_window)};
+    auto& limiter{*Assert(limiter_)};
```
Personally, I would rather vote for simpler final code than a bit more involved diff

l0rinc reviewed (approved):
Code review ACK 5c74a0b397cb3db94761bad78801eed4544155b9

I still have a few suggestion that I think should be addressed, but I'm not blocking.

Crypt-iQ commented on src/logging.cpp:
```diff
@@ -455,24 +455,26 @@ void BCLog::Logger::LogPrintStr_(std::string_view str, std::source_location&& so
-        }
+    }
+
+    // To avoid confusion caused by dropped log messages when debugging an issue,
+    // we prefix log lines with "[*]" when there are any suppressed source locations.
```
This refers to situations that could pop up like if functional tests were failing and the logs were being suppressed. Or even somebody having an issue with a production node.

achow101 commented:
ACK 5c74a0b397cb3db94761bad78801eed4544155b9

fanquake commented:
This might be causing sporadic CI failures? See #33195.
